File: ict_trading_bot/utils/mt5_credentials.py
import logging
import os

from supabase import create_client

logger = logging.getLogger(__name__)


def fetch_mt5_credentials():
    url = os.getenv("SUPABASE_URL")
    key = os.getenv("SUPABASE_KEY")

    if not url or not key:
        raise RuntimeError("SUPABASE_URL and SUPABASE_KEY are required to load MT5 credentials")

    client = create_client(url, key)

    try:
        res = (
            client.table("mt5_credentials")
            .select("login,password,server,updated_at")
            .eq("active", True)
            .order("updated_at", desc=True)
            .limit(1)
            .execute()
        )
    except Exception as exc:
        raise RuntimeError(f"Failed to fetch MT5 credentials from Supabase: {exc}") from exc

    data = getattr(res, "data", None) or []
    if not data:
        raise RuntimeError("No MT5 credentials found in Supabase (mt5_credentials)")

    row = data[0]
    return {
        "login": row.get("login"),
        "password": row.get("password"),
        "server": row.get("server"),
    }

File: jaguar-main/pages/api/admin/mt5-credentials.js
import { createServerSupabaseClient } from "@supabase/auth-helpers-nextjs";
import { getSupabaseClient } from "../../../lib/supabaseClient";

export default async function handler(req, res) {
  if (req.method !== "GET" && req.method !== "POST") {
    return res.status(405).end();
  }

  try {
    const supabase = createServerSupabaseClient({ req, res });
    const {
      data: { session },
    } = await supabase.auth.getSession();

    if (!session || !session.user) {
      return res.status(401).json({ error: "not authenticated" });
    }

    const supabaseAdmin = getSupabaseClient({ server: true });
    const userId = session.user.id;
    const { data: profile } = await supabaseAdmin
      .from("profiles")
      .select("role")
      .eq("id", userId)
      .maybeSingle();

    const role = (profile?.role || "user").toLowerCase();
    if (role !== "admin") {
      return res.status(403).json({ error: "forbidden" });
    }

    if (req.method === "GET") {
      const { data, error } = await supabaseAdmin
        .from("mt5_credentials")
        .select("login, server, updated_at")
        .eq("active", true)
        .order("updated_at", { ascending: false })
        .limit(1);

      if (error) {
        return res.status(500).json({ error: "failed to load credentials" });
      }

      const row = Array.isArray(data) ? data[0] : null;
      return res.status(200).json({
        credentials: row
          ? {
              login: row.login || "",
              server: row.server || "",
              updated_at: row.updated_at || null,
              hasPassword: true,
            }
          : {
              login: "",
              server: "",
              updated_at: null,
              hasPassword: false,
            },
      });
    }

    const { login, password, server } = req.body || {};
    if (!login || !password || !server) {
      return res.status(400).json({ error: "login, password, and server are required" });
    }

    await supabaseAdmin.from("mt5_credentials").update({ active: false }).eq("active", true);

    const { error: insertError } = await supabaseAdmin
      .from("mt5_credentials")
      .insert({
        login: String(login).trim(),
        password: String(password),
        server: String(server).trim(),
        active: true,
        updated_at: new Date().toISOString(),
      });

    if (insertError) {
      return res.status(500).json({ error: "failed to save credentials" });
    }

    return res.status(200).json({ ok: true });
  } catch (e) {
    console.error(e);
    return res.status(500).json({ error: e.message || String(e) });
  }
}
